cubic-server/recipes/CraftingShaped.cpp

```diff
@@ -0,0 +1,127 @@
+            !recipe["result"].is_object() || \
+            !recipe["result"].contains("item") || \
+            !recipe["result"]["item"].is_string())
+            return;
+        this->_result = Server::getInstance()->getItemConverter().fromItemToProtocolId(recipe["result"]["item"].get<std::string>());
+        if (recipe["result"].contains("count") && recipe["result"]["count"].is_number_unsigned())
+            this->_count = recipe["result"]["count"].get<nlohmann::json::number_unsigned_t>();
+        else
+            this->_count = 1;
+        for (size_t pattern_line_pos = 0; pattern_line_pos < recipe["pattern"].size(); pattern_line_pos++) {
+            if (!recipe["pattern"][pattern_line_pos].is_string())
+                return;
+            this->_pattern.push_back(std::vector<MaybeItems>());
+
+            std::string pattern_line = recipe["pattern"][pattern_line_pos].get<std::string>();
+
+            for (const auto &pattern_item : pattern_line) {
+                if (pattern_item == ' ')
+                    this->_pattern[pattern_line_pos].push_back(MaybeItems());
+                else {
+                    if (!recipe["key"].contains(&pattern_item))
+                        return;
+                    if (!this->getKey(pattern_item, recipe["key"][&pattern_item]))
+                        return;
+                    this->_pattern[pattern_line_pos].push_back(MaybeItems(this->_key[pattern_item]));
+                }
+            }
+        }
+        this->setValidity(true);
+    }
+
+    void CraftingShaped::dump(void) const
+    {
+        for (const auto &line : this->_pattern) {
+            for (const auto &item : line) {
+                if (item.isEmpty())
+                    std::cout << ' ';
+                else {
+                    bool found_key = false;
+
+                    for (const auto &[key, items] : this->_key) {
+                        if (item.getItems().lock() == items) {
+                            std::cout << key;
+                            found_key = true;
+                            break;
+                        }
+                    }
+                    if (!found_key)
+                        std::cout << '!';
+                }
+            }
+            std::cout << std::endl;
+        }
+        for (const auto &[key, items] : this->_key) {
+            std::cout << '\''<< key << "\': [";
+            bool first_item = true;
+
+            for (const auto &item : *items) {
+                if (first_item)
+                    first_item = false;
+                else
+                    std::cout << ", ";
+                std::cout << '\"' << Server::getInstance()->getItemConverter().fromProtocolIdToItem(item) << '\"';
+            }
+            std::cout << ']' << std::endl;
+        }
+        std::cout << " -> " << Server::getInstance()->getItemConverter().fromProtocolIdToItem(this->_result) << " (x" << this->_count << ')' << std::endl;
```

Comment:
please use a string or a stringstream for all the std::cout
Same as Blasting.cpp